### ISSUE TITLE
Unable to add unique expression index on REGIONAL BY ROW table

### ISSUE THREAD

**Issue by smcvey** [cockroachdb/cockroach#83076]
Reproduction steps:

* Build a 3 node cluster with localities 'central', 'east' and 'west' and then...

```
CREATE DATABASE my_db PRIMARY REGION central REGIONS west, east;
USE my_db;
CREATE TABLE users (
   id UUID PRIMARY KEY DEFAULT gen_random_uuid(),                                                                                                                                    
   email STRING NOT NULL
);
ALTER TABLE users SET LOCALITY REGIONAL BY ROW;
```

Then trying this fails:

```
root@:26257/my_db> CREATE UNIQUE INDEX on users (LOWER(email));
ERROR: validate unique constraint: column "crdb_internal_idx_expr" does not exist
SQLSTATE: 42703CREATE UNIQUE INDEX on users (LOWER(email));
ERROR: validate unique constraint: column "crdb_internal_idx_expr" does not exist
SQLSTATE: 42703
```

It works fine as a REGIONAL BY TABLE, just not REGIONAL BY ROW.

However, if I leave it as REGIONAL BY TABLE, add the unique index and THEN make it regional by row, that works.

I am able to work around this by switching the table to REGIONAL BY TABLE, add the unique index and then switch it back, but this is not ideal.

Jira issue: CRDB-16846

**Comment by smcvey** [comment]
Affects 22.1 and 21.2.

**Comment by smcvey** [comment]
Removing the `lower()` means it works also:

```
root@:26257/my_db> CREATE UNIQUE INDEX on users (email);
CREATE INDEX
```

**Comment by Xiang-Gu** [comment]
Giving it a stab now. I am familiarizing myself with the workflow of creating an expression index. The `crdb_internal_idx_expr` column is the "expression" column (computed, virtual, inaccessible) created for the expression index. 

I'm now trying to figure out why such a column is not created with a REGIONAL BY ROW table.

**Comment by Xiang-Gu** [comment]
After some digging, I think the issue is that after we merged the newly created index and the temporary index, we went ahead and validate it, as in
https://github.com/cockroachdb/cockroach/blob/73383399b94f46ac70ef286cb10a170f277e3235/pkg/sql/backfill.go#L1424-L1436

in which we attempted to count the number of rows and uniqueness of this newly created index, as in 
https://github.com/cockroachdb/cockroach/blob/73383399b94f46ac70ef286cb10a170f277e3235/pkg/sql/backfill.go#L1687

in which we attempted to check uniqueness, as in
https://github.com/cockroachdb/cockroach/blob/73383399b94f46ac70ef286cb10a170f277e3235/pkg/sql/backfill.go#L1893-L1908 

With a REGIONAL BY ROW table, the `idx.GetPartitioning().NumImplicitColumns() > 0` will be true and hence we entered this block to check for uniqueness on the `crdb_internal_idx_expr` column. This column is inaccessible so the query inside `validateUniqueConstraint` will fail and error out "crdb_internal_idx_expr does not exist".

**Comment by Xiang-Gu** [comment]
It's tempting to just add a clause there and say "if it's a virtual column, don't enter and check uniqueness", but I don't know why we would have this uniqueness check here in the first place. This is not to validate, in general, the uniqueness of a to-be-created unique index, since that's done earlier (I used a debuger and it errors out with message 
```
"pq: failed to ingest index entries during backfill: duplicate key value violates unique constraint "idx" 
```
before reaching this block of code; I cannot pinpoint the exact location tho).

**Comment by stevendanna** [comment]
> Can someone, maybe https://github.com/orgs/cockroachdb/teams/bulk-io , help explain to me in what case is this uniqueness check needed

I'm not an expert in the schema side of things, but I can try.

> It's tempting to just add a clause there and say "if it's a virtual column, don't enter and check uniqueness", but I don't know why we would have this uniqueness check here in the first place. This is not to validate, in general, the uniqueness of a to-be-created unique index, since that's done earlier
>
> "pq: failed to ingest index entries during backfill: duplicate key value violates unique constraint "idx" 

I'm guessing that to get this error you created a relatively small table with a duplicate in the column you planned to make a unique index on.  In this case, you _may_ get a uniqueness violation during the backfill process but it isn't guaranteed. To catch cases where it isn't guaranteed, we need to recheck at the end.

It isn't guaranteed to occur during the backfill for a few reasons.  First, the bulk ingestion process uses AddSSTable which takes an SST that we construct based on the existing table and sends it to pebble, rather than the typical write path which uses a conditional put to report on duplicate key violations. You happen to get a duplicate key violation during the backfill because we have some code in the SSTBatcher (the component that constructs SSTs to send to AddSSTable), that detects duplicates _within_ an SST before we send it.  But if we change some cluster settings to force our duplicate key into a different SST, we'll see we get a slightly different error:

```
set cluster setting schemachanger.backfiller.buffer_size='64';
set cluster setting schemachanger.backfiller.buffer_size='64'; 
create unique index on foo(j);
ERROR: duplicate key value violates unique constraint "foo_j_key"
SQLSTATE: 23505
CONSTRAINT: foo_j_key
```

We get the slightly different error because in this case the duplicate wasn't caught until our check at the end.  Another reason you can get duplicates is because we don't use CPuts during the merging process and thus further duplicates could sneak in there.

Now, there are actually two uniqueness checks happening here: one based on counting rows and another based on a query to find duplicates.

In the first, we count all the rows in the index and compare that to the count we are supposed to have. If the index is unique, it should have the same number of rows as the primary index, which is also unique, adjusting appropriately for partial indexes. If there is a count mismatch, we report a duplicate since presumably duplicate values led to writes to the same key in the index, leading to fewer values than we expect.

The second check based on the query to find duplicates is the one you are asking about here:

> "for implicitly partitioned unique indexes, we need to independently validate that the non-implicitly partitioned columns are unique"? That is, what is an example that, we don't have uniqueness on the explicit columns, but it still passes the earlier uniqueness check due to the presence of an implicitly partitioning?

My best understanding of why this is needed is that our counting strategy has a problem for implicitly partitioned indexes. The problem is that an implicitly partitioned index has columns in the index key that could lead our counts to match even in the case of a duplicate value. For example, in a regional by row table, a new column (`crdb_region`) is made part of the key for our primary index.  So, say we had a table like:

```
CREATE TABLE foo (
   id int PRIMARY KEY,
   b int
);
ALTER TABLE foo SET LOCALITY REGIONAL BY ROW;
```

Now say we add two conflicting rows values (e.g. `(1, 1)` and `(2, 1)`) but one was added in the west region and one in the east.  If we made a unique index on `foo(b)`, our backfill will write out two keys despite of the duplicate because the region is in the key:

    /Table/113/2/"east"/1
    /Table/113/2/"west"/1

and our counting strategy would detect no duplicate.  However, the check in `validateUniqueConstraint` would. And in fact, we can sort of see that it is this check that saves us because we get yet another variation on our duplicate constraint error:

```
ERROR: could not create unique constraint "foo_b_key"              
SQLSTATE: 23505                                                                                   
DETAIL: Key (b)=(1) is duplicated.                                                                                                                                                                  
CONSTRAINT: foo_b_key 
```

To summarize:

- We need uniqueness checks after the backfill because the backfill process writes to the index in ways that bypass the uniqueness checks that typically occur on the write path.
- We can't rely simply on counting for implicitly partitioned indexes because implicitly partitioned indexes include columns in their key that may give duplicate values unique keys in the index.

I'm not an expert here, but I hope that helps point you in the right direction. Hopefully it doesn't include anything too untrue.

**Comment by stevendanna** [comment]
> It's tempting to just add a clause there and say "if it's a virtual column, don't enter and check uniqueness", but I don't know why we would have this uniqueness check here in the first place. 

I don't think this would be correct.  In the case that shaun posted, I think it is clear the user _wants_ to catch duplicates that would be caused by the emails "Foo@example.com" and "foo@example.com".  I think we'll need to fix our uniqueness check query to support expressions.
